### PR TITLE
Only reuse memory in mapslices when it's certainly safe (fixes #18524)

### LINKED ISSUE #18524: Trouble from reusing temporary in mapslices
Ref https://groups.google.com/d/msg/julia-users/0im3DGRNbG0/_oH0TZ1TAgAJ

``` julia
julia> mapslices(x->tuple(copy(x)), [1 2; 3 4], 1)
1×2 Array{Tuple{Array{Int64,1}},2}:
 ([1,3],)  ([2,4],)

julia> mapslices(x->tuple(x), [1 2; 3 4], 1)
1×2 Array{Tuple{Array{Int64,1}},2}:
 ([2,4],)  ([2,4],)
```

Is there any alternative to going back to allocating a new array for each slice, even when it shouldn't be necessary?

### PR DESCRIPTION
**Pull request by timholy**
See discussion in #18524. This avoids too much pessimization and is hopefully safe.

### CODE REVIEW

**Comment by simonster** [comment]
Fine with me and probably fine for all existing `mapslices` calls, but I still don't think this would produce the previous result for a pathological case like `mapslices(x->[x, x], A, 1)`. Should also adjust the note in NEWS.md and perhaps document this behavior explicitly.

**Comment by timholy** [comment]
Fair enough. We could presumably introspect and see whether the result aliases `Aslice`, but that seems like it could be expensive.

**Comment by timholy** [comment]
OK, this should be even safer.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -1737,6 +1737,7 @@ function mapslices(f, A::AbstractArray, dims::AbstractVector)
 
     Aslice = A[idx...]
     r1 = f(Aslice)
+    safe_for_reuse = isa(r1, Number) || (isa(r1, AbstractArray) && eltype(r1) <: Number)
```
Comment (tkelman) [inline base/abstractarray.jl]: what is special about `Number` ?

Comment (timholy) [inline base/abstractarray.jl]: immutable and doesn't contain anything (and happens to cover the vast majority of what people use `mapslices` in practice for).

Comment (tkelman) [inline base/abstractarray.jl]: will it be safe even for things like BigFloat ?

Comment (simonster) [inline base/abstractarray.jl]: Yes, it's "safe" for any object that couldn't possibly contain the passed array.

**Comment by simonster** [comment]
If you want to write really pathological code, you could still do this:

```julia
arrays = []
mapslices(x->(push!(arrays, x); 0), A)
```

I don't think anyone is actually going to do this, but I am a little concerned of the complexity of actually documenting why you can't.

File: base/abstractarray.jl (hunk 1 above)
Comment (StefanKarpinski) [inline base/abstractarray.jl]: probably safe for strings as well

**Comment by StefanKarpinski** [comment]
I think this just needs a rebase and merge.

**Comment by tkelman** [comment]
Bump?

**Comment by tkelman** [comment]
I just tried the new github "resolve conflicts" feature, double-check that I didn't mess it up too badly

**Pull request merged**